lib/qmp.ml

```diff
@@ -99,126 +99,156 @@ module Event = struct
+      let family  = json |> U.member "family"  |> U.to_string in
+      let service = json |> U.member "service" |> U.to_string |> int_of_string in
+      let host    = json |> U.member "host"    |> U.to_string in
+      { enabled; auth; family; service; host}
+    in
+    let xen_platform_pv_driver_info json =
+      let product_num = json |> U.member "product-num" |> U.to_int in
+      let build_num   = json |> U.member "build-num"   |> U.to_int in
+      {product_num; build_num}
+    in
+    let query_kvm json =
+      let enabled = json |> U.member "enabled" |> U.to_bool in
+      let present = json |> U.member "present" |> U.to_bool in
+      {enabled; present}
+    in
+    let name_list json =
+      json |> U.to_list |> List.map (fun x -> x |> U.member "name" |> U.to_string)
+    in
+    let add_fd json =
+      let fd       = json |> U.member "fd"       |> U.to_int in
+      let fdset_id = json |> U.member "fdset-id" |> U.to_int in
+      {fd; fdset_id}
+    in
+    let result =
+      let return = json |> U.member "return" in
+      return |> function
+      | `List  _ -> return |> name_list |> fun x -> Name_list x
+      | `Assoc _ -> (return |> U.keys |> function
+        | []         -> Unit
+        | x when x |> for_all ["status"]                   -> return |> status    |> fun x -> Status x
+        | x when x |> for_all ["enabled"; "auth"; "family"; "service"; "host"] -> return |> query_vnc |> fun x -> Vnc x
+        | x when x |> for_all ["product-num"; "build-num"] -> return |> xen_platform_pv_driver_info   |> fun x -> Xen_platform_pv_driver_info x
+        | x when x |> for_all ["enabled"; "present"]       -> return |> query_kvm |> fun x -> Enabled x
+        | x when x |> for_all ["fd"; "fdset-id"]           -> return |> add_fd    |> fun x -> Fd_info x
+        | _ -> failwith (Printf.sprintf "unknown result %s" (Y.to_string return))
         )
-      | x -> failwith (Printf.sprintf "unknown result %s" (Yojson.Safe.to_string x))
-    )
-  | `Assoc list when List.mem_assoc "error" list ->
-    let id = if List.mem_assoc "id" list then Some (string (List.assoc "id" list)) else None in
-    let error = assoc (List.assoc "error" list) in
-    let cls = string (List.assoc "class" error) in
-    let descr = string (List.assoc "desc" error) in
-    Error (id, {cls; descr})
-  | x ->
-    failwith (Printf.sprintf "message_of_string %s" (Yojson.Safe.to_string x))
+      | _ -> failwith (Printf.sprintf "unknown result type %s" (Y.to_string json))
+    in
+    (json |> id, result)
+  in
+
+  let error json =
+    let cls   = json |> U.member "error" |> U.member "class" |> U.to_string in
+    let descr = json |> U.member "error" |> U.member "desc"  |> U.to_string in
+    ( json |> id, {cls; descr})
+  in
+
+  let parse_err descr = Error (None, { cls = _JSONParsing; descr }) in
+
+  try
+    let json = Y.from_string str in
+    match json |> U.keys with
+    | x when x |> for_all ["QMP"]     -> json |> greeting |> fun x -> Greeting x
```

Comment:
Can there ever be more than one key in a legal message? I would have matched that one key directly:
```
match json |> U.keys with
| ["QMP"] -> ..
| ["event"] -> .. 
```